nav2_map_server/src/map_saver.cpp

```diff
@@ -30,221 +31,173 @@
 
 #include "nav2_map_server/map_saver.hpp"
 
-#include <cstdio>
-#include <fstream>
-#include <memory>
-#include <string>
-#include <utility>
-#include <vector>
-
-#include "Magick++.h"
-#include "nav2_map_server/map_mode.hpp"
-#include "nav_msgs/msg/occupancy_grid.h"
-#include "nav_msgs/srv/get_map.hpp"
-#include "rclcpp/rclcpp.hpp"
-#include "tf2/LinearMath/Matrix3x3.h"
-#include "tf2/LinearMath/Quaternion.h"
-#include "yaml-cpp/yaml.h"
+#include <stdexcept>
 
 namespace nav2_map_server
 {
-MapSaver::MapSaver(const rclcpp::NodeOptions & options)
-: Node("map_saver", options), save_next_map_promise{}
+MapSaver::MapSaver()
+: nav2_util::LifecycleNode("map_saver")
 {
-  Magick::InitializeMagick(nullptr);
-  {
-    mapname_ = declare_parameter("output_file_no_ext", "map");
-    if (mapname_ == "map") {
-      mapname_ += "_" + std::to_string(static_cast<int>(now().seconds()));
-    }
-    threshold_occupied_ = declare_parameter("threshold_occupied", 65);
-    if (100 < threshold_occupied_) {
-      throw std::runtime_error("Threshold_occupied must be 100 or less");
-    }
-    threshold_free_ = declare_parameter("threshold_free", 25);
-    if (threshold_free_ < 0) {
-      throw std::runtime_error("Free threshold must be 0 or greater");
-    }
-    if (threshold_occupied_ <= threshold_free_) {
-      throw std::runtime_error("Threshold_free must be smaller than threshold_occupied");
-    }
+  RCLCPP_INFO(get_logger(), "Creating");
 
-    std::string mode_str = declare_parameter("map_mode", "trinary");
-    try {
-      map_mode = map_mode_from_string(mode_str);
-    } catch (std::invalid_argument &) {
-      map_mode = MapMode::Trinary;
-      RCLCPP_WARN(
-        get_logger(), "Map mode parameter not recognized: '%s', using default value (trinary)",
-        mode_str.c_str());
-    }
+  save_map_timeout_ = std::make_shared<rclcpp::Duration>(
+    std::chrono::milliseconds(declare_parameter("save_map_timeout", 2000)));
 
-    image_format = declare_parameter("image_format", map_mode == MapMode::Scale ? "png" : "pgm");
-    std::transform(
-      image_format.begin(), image_format.end(), image_format.begin(),
-      [](unsigned char c) {return std::tolower(c);});
-    const std::vector<std::string> BLESSED_FORMATS{"bmp", "pgm", "png"};
-    if (
-      std::find(BLESSED_FORMATS.begin(), BLESSED_FORMATS.end(), image_format) ==
-      BLESSED_FORMATS.end())
-    {
-      std::stringstream ss;
-      bool first = true;
-      for (auto & format_name : BLESSED_FORMATS) {
-        if (!first) {
-          ss << ", ";
+  free_thresh_default_ = declare_parameter("free_thresh_default", 25),
```

Comment:
Should these be int or double? (real question, I'm not sure)

Reply:
Currently for map loading logic in `LoadParameters` structure it is used double type scaled from [0.0..1.0] for keeping the thresholds. Also, floating-point is written to `map.yaml` files. In map saving logic it also finally transformed to floating-point before writing to a yaml: https://github.com/ros-planning/navigation2/blob/master/nav2_map_server/src/map_saver.cpp#L223. For unification I may change thresholds for `SaveParameters` to double as well. But I'd like to prefer to remain them untouched: [0..100] percent integer range for map saving logic because it is more convenient to use when calling services manually from `rqt` or console. Probably switch `SaveParameters` to using double thresholds, but remain int percent scale for services API?

Reply:
I leave it up to you to do what you think is best, I'm just bringing it up as a question.

Reply:
Map data is stored in `int8` and all map thresholds checks are using integer values . However, `map*.yaml` files already use double values. So, in order to not break the compatibility with older yaml-s and for API unification, I am going to make all threshold to be double in next commit (even in map saving services).

Reply:
OK works for me

Reply:
Done: https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-b3fa8804c8edd37f085c3ee28e9a092bR87